### PR TITLE
Refactor trello card

### PR DESCRIPTION
**Pull request by wokenqingtian**
#### Refactor `Trello::Card#initialize_fields` to be more readable and get ready for future further refactoring
 
Break codebase into three parts:
1. Parse and update read-only attributes
2. Parse and update writable attributes
3. Serialize or format some attributes

#### Reduce the responsibilities of `Trello::Card#update_fields`

Previous `update_fields` method has two responsibilities:

1. For the gem user to update fields
2. For JsonUtils to prase response

Now: 
1. I remove the option 2 from it. And refactor `update_fields` to only update writable attributes
2. Then make JsonUtils direct use `initialize_fields` method.

### CODE REVIEW

File: lib/trello/card.rb [hunk 1]
```diff
@@ -491,31 +481,49 @@ def fields_has_key?(fields, key)
...
+      if fields.key?(api_version_key)
+        fields[api_version_key]
+      elsif fields.key?(gem_version_key)
+        fields[gem_version_key]
+      else
+        attributes[gem_version_key]
+      end
+    end
+
+    def parse_readonly_fields(fields, key)
```
Comment (runlevel5) [inline lib/trello/card.rb:518]: What do you think about setting a default value for `fields` param?

File: lib/trello/card.rb [hunk 2]
```diff
@@ -491,31 +481,49 @@ def fields_has_key?(fields, key)
...
+
+      %i[
+        name desc due due_complete closed board_id
+        member_ids list_id pos labels card_labels
+        cover_image_id source_card_id source_card_properties
+      ].each do |attr_key|
+        attributes[attr_key] = parse_writable_fields(fields, attr_key)
+      end
+
+      attributes[:last_activity_date] = Time.iso8601(attributes[:last_activity_date]) rescue nil if attributes[:last_activity_date]
```
Comment (runlevel5) [inline lib/trello/card.rb:499]: it is a bit hard to read the time casting, what do you think about having a method for that responsibility?

Comment (runlevel5) [reply lib/trello/card.rb:499]: Plus, what do you think about removing the presence check for `attributes[:last_activity_date]`? Isn't that `rescue nil` would handle the case `Time.iso8601(nil)`?

File: lib/trello/card.rb [hunk 3]
```diff
@@ -491,31 +481,49 @@ def fields_has_key?(fields, key)
...
 
+    def parse_writable_fields(fields, key)
+      gem_version_key = key.to_sym
+      api_version_key = SYMBOL_TO_STRING[gem_version_key]
+      if fields.key?(api_version_key)
+        fields[api_version_key]
+      elsif fields.key?(gem_version_key)
+        fields[gem_version_key]
+      else
+        attributes[gem_version_key]
```
Comment (runlevel5) [inline lib/trello/card.rb:514]: Can we refactor to `fields[api_version_key] || fields[gem_version_key] || attributes[gem_version_key]`?

File: lib/trello/card.rb [hunk 4]
```diff
@@ -491,31 +481,49 @@ def fields_has_key?(fields, key)
...
+    end
+
+    def parse_readonly_fields(fields, key)
+      gem_version_key = key.to_sym
+      api_version_key = SYMBOL_TO_STRING[gem_version_key]
+      if fields.key?(api_version_key)
+        fields[api_version_key]
+      else
+        attributes[gem_version_key]
+      end
```
Comment (runlevel5) [inline lib/trello/card.rb:525]: What do you think about refactoring those lines to `fields[api_version_key] || attributes[gem_version_key]`?

Comment (wokenqingtian) [reply lib/trello/card.rb:525]: Yes, you're right, I should do this for readonly attributes.

File: lib/trello/card.rb (hunk 3 above)
Comment (wokenqingtian) [reply lib/trello/card.rb:514]: There's a circumstance maybe we should consider. If I want to set an attribute to nil, the "||" may get a wrong result.

```ruby
card = Trello::Card.new(name: 'card name')
card.update_fields(name: nil)
card.name #=> get "card name" but not nil
```

File: lib/trello/card.rb (hunk 2 above)
Comment (wokenqingtian) [reply lib/trello/card.rb:499]: I will do this

File: lib/trello/card.rb (hunk 1 above)
Comment (wokenqingtian) [reply lib/trello/card.rb:518]: Maybe I can set it here?

```ruby
def initialize_fields(fields)
  fields ||= {}
```

Is that ok?

File: lib/trello/card.rb (hunk 3 above)
Comment (runlevel5) [reply lib/trello/card.rb:514]: I see. Please disregard my comment :)

File: lib/trello/card.rb (hunk 1 above)
Comment (wokenqingtian) [reply lib/trello/card.rb:520]: I added a default value for fields in both `parse_writable_fields` and `parse_readonly_fields` methods

File: lib/trello/card.rb [hunk 5]
```diff
@@ -516,14 +518,26 @@ def parse_writable_fields(fields, key)
...
-        fields[api_version_key]
-      else
-        attributes[gem_version_key]
-      end
+
+      fields[api_version_key] || attributes[gem_version_key]
     end
 
+    def serialize_time(time)
+      return time unless time.is_a?(String)
```
Comment (runlevel5) [inline lib/trello/card.rb:529]: What could `time` be in this case? Should we return `time` if it is of a DateTime object?

Comment (wokenqingtian) [reply lib/trello/card.rb:529]: After some research, I think we should only support Time, DateTime, String only.

Should we write defensive code to check the object's class like this?

```ruby
return nil unless time.is_a?(String) || time.is_a?(Time) || time.is_a?(DateTime)
# or
raise ArgumentError unless time.is_a?(String) || time.is_a?(Time) || time.is_a?(DateTime)
```
Or just keep current feature only check the string?

### URL encode format difference between Time and DateTime:
Rest-client use `URI.encode_www_form_component` to format params.

```bash
t1 =Time.new(2019, 1, 1, 1, 1, 1, '+00:00')
t2 =DateTime.new(2019,1,1,1,1,1, '+00:00')

URI.encode_www_form_component(t1) #=> 2019-01-01+01%3A01%3A01+%2B0000
URI.encode_www_form_component(t2) #=> 2019-01-01T01%3A01%3A01%2B00%3A00
```

I also test both of this formats with Trello API. All works fine. So I think we can leave this difference here. In future, it's better to pre-format this field to a unified format before submit the request.

Comment (runlevel5) [reply lib/trello/card.rb:529]: > I also test both of this formats with Trello API. All works fine. So I think we can leave this difference here. In future, it's better to pre-format this field to a unified format before submit the request.

:+1:

**Pull request merged**